compiler/src/dotty/tools/dotc/typer/Namer.scala

```diff
@@ -1128,7 +1128,8 @@ class Namer { typer: Typer =>
       // it would be erased to BoxedUnit.
       def dealiasIfUnit(tp: Type) = if (tp.isRef(defn.UnitClass)) defn.UnitType else tp
 
-      val rhsCtx = ctx.addMode(Mode.InferringReturnType)
+      var rhsCtx = ctx.addMode(Mode.InferringReturnType)
+      if (sym.isTransparentMethod) rhsCtx = rhsCtx.addMode(Mode.TransparentBody)
```

Comment:
I'm not sure `TransparentBody` should be a Mode. If it is, shouldn't it also be set in the other tree traversing helpers besides typer, i.e., `TreeAccumulator` and `MegaPhase#transformTree`? All of those have `localContext` helpers (in case of typer its in `NamerContextOps`, the other two are local defs) which currently set the owner, among other things.

As an alternative, we could deduce whether we are in a transparent context by walking the context's owner chain and checking for `Transparent`. Ideally that query would be cached by the context (and its descendants).

Reply:
Good point. Since it's a single bit we cache, it could still go into the `Mode` structure (context fields are a precious resource). But it could be maintained fully automatically. I'll do that change.